File: hphp/hack/src/hhbc/emit_constant.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
open Instruction_sequence
module T = Aast

let emit_constant_cinit env (ast_constant : Tast.gconst) (c : Hhas_constant.t) :
    Hhas_function.t option =
  let (ns, name) =
    Hhas_constant.name c
    |> Hhbc_id.Const.from_ast_name
    |> Hhbc_id.Const.to_raw_string
    |> Utils.split_ns_from_name
  in
  let name = ns ^ "86cinit_" ^ name in
  let original_id = Hhbc_id.Function.from_ast_name name in
  let ret = ast_constant.T.cst_type in
  let return_type_info =
    match ret with
    | None -> None
    | Some h ->
      Some
        (Emit_type_hint.hint_to_type_info
           ~kind:Emit_type_hint.Return
           ~skipawaitable:false
           ~nullable:false
           ~tparams:[]
           h)
  in
  (* let return_type_info = ast_constant.T.cst_type in *)
  let instrs = Hhas_constant.initializer_instrs c in
  match instrs with
  | None -> None
  | Some instrs ->
    let verify_instr =
      match return_type_info with
      | None -> Instr_empty
      | Some _ -> instr_verifyRetTypeC
    in
    let instrs = gather [instrs; verify_instr; instr_retc] in
    let body =
      Emit_body.make_body
        instrs
        [] (* decl_vars *)
        false (* is_memoize_wrapper *)
        false (* is_memoize_wrapper_lsb *)
        [] (* upper_bounds *)
        [] (* shadowed_tparams *)
        [] (* params *)
        return_type_info
        None (* doc_comment *)
        (Some env)
    in
    Some
      (Hhas_function.make
         [] (* attributes *)
         original_id
         body
         (Hhas_pos.pos_to_span ast_constant.T.cst_span)
         false (* is_async *)
         false (* is_generator *)
         false (* is_pair_generator *)
         Closure_convert.TopLevel
         true (* no_injection *)
         false (* is_interceptable *)
         false (* is_memoize_impl *)
         Rx.NonRx
         false (* rx_disabled *))

let emit_constant env (ast_constant : Tast.gconst) :
    Hhas_constant.t * Hhas_function.t option =
  let c =
    Hhas_constant.from_ast
      env
      ast_constant.T.cst_name
      (Some ast_constant.T.cst_value)
  in
  let f = emit_constant_cinit env ast_constant c in
  (c, f)

let emit_constants_from_program env (ast : Tast.def list) :
    Hhas_constant.t list * Hhas_function.t list =
  let aux def =
    match def with
    | T.Constant c -> Some (emit_constant env c)
    | _ -> None
  in
  let constant_tuples = List.filter_map ast ~f:aux in
  let (constants, constant_cinits_optionals) = List.unzip constant_tuples in
  let constant_cinits = List.filter_opt constant_cinits_optionals in
  (constants, constant_cinits)

File: hphp/hack/src/hhbc/emit_constant.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use core_utils_rust as utils;
use emit_body_rust as emit_body;
use emit_type_hint_rust::{self as emit_type_hint, Kind};
use env::{emitter::Emitter, Env};
use hhas_coeffects::HhasCoeffects;
use hhas_constant_rust::{self as hhas_constant, HhasConstant};
use hhas_function_rust::{self as hhas_function, HhasFunction};
use hhas_pos_rust::Span;
use hhbc_id_rust::{r#const, function, Id};
use hhbc_string_utils_rust::strip_global_ns;
use instruction_sequence::{instr, InstrSeq, Result};
use oxidized::ast as tast;

fn emit_constant_cinit<'a>(
    e: &mut Emitter,
    env: &mut Env<'a>,
    constant: &'a tast::Gconst,
    c: &HhasConstant<'a>,
) -> Result<Option<HhasFunction<'a>>> {
    let const_id = r#const::Type::from_ast_name(&constant.name.1);
    let (ns, name) = utils::split_ns_from_name(const_id.to_raw_string());
    let name = String::new() + strip_global_ns(ns) + "86cinit_" + name;
    let original_id: function::Type = name.into();
    let ret = constant.type_.as_ref();
    let return_type_info = ret
        .map(|h| {
            emit_type_hint::hint_to_type_info(
                &Kind::Return,
                false, /* skipawaitable */
                false, /* nullable */
                &[],   /* tparams */
                h,
            )
        })
        .transpose()?;
    c.initializer_instrs
        .as_ref()
        .map(|instrs| {
            let verify_instr = match return_type_info {
                None => instr::empty(),
                Some(_) => instr::verify_ret_type_c(),
            };
            let instrs = InstrSeq::gather(vec![instrs.clone(), verify_instr, instr::retc()]);
            let body = emit_body::make_body(
                e,
                instrs,
                vec![],
                false,  /* is_memoize_wrapper */
                false,  /* is_memoize_wrapper_lsb */
                vec![], /* upper_bounds */
                vec![], /* shadowed_params */
                vec![], /* params */
                return_type_info,
                None, /* doc_comment */
                Some(&env),
            )?;
            Ok(HhasFunction {
                attributes: vec![],
                name: original_id.into(),
                body,
                span: Span::from_pos(&constant.span),
                coeffects: HhasCoeffects::default(),
                flags: hhas_function::Flags::NO_INJECTION,
            })
        })
        .transpose()
}

fn emit_constant<'a>(
    e: &mut Emitter,
    env: &mut Env<'a>,
    constant: &'a tast::Gconst,
) -> Result<(HhasConstant<'a>, Option<HhasFunction<'a>>)> {
    let c = hhas_constant::from_ast(e, env, &constant.name, Some(&constant.value))?;
    let f = emit_constant_cinit(e, env, constant, &c)?;
    Ok((c, f))
}

pub fn emit_constants_from_program<'a>(
    e: &mut Emitter,
    env: &mut Env<'a>,
    defs: &'a [tast::Def],
) -> Result<(Vec<HhasConstant<'a>>, Vec<HhasFunction<'a>>)> {
    let const_tuples = defs
        .iter()
        .filter_map(|d| d.as_constant().map(|c| emit_constant(e, env, c)))
        .collect::<Result<Vec<_>>>()?;
    let (contants, inits): (Vec<_>, Vec<_>) = const_tuples.into_iter().unzip();
    Ok((contants, inits.into_iter().filter_map(|x| x).collect()))
}
